agent/fw_laravel_queue.c

```diff
@@ -40,7 +39,324 @@ static int nr_laravel_queue_is_sync_job(zval* job TSRMLS_DC) {
+    const char* key,
+    const nrobj_t* val,
+    nr_laravel_queue_headers_t* headers) {
+  char* key_lc;
+
+  if (NULL == headers) {
+    return NR_SUCCESS;
+  }
+
+  key_lc = nr_string_to_lowercase(key);
+  if (NULL == key_lc) {
+    return NR_SUCCESS;
+  }
+
+  if (0 == nr_strcmp(key_lc, X_NEWRELIC_ID_MQ_LOWERCASE)) {
+    headers->id = nro_get_string(val, NULL);
+  } else if (0 == nr_strcmp(key_lc, X_NEWRELIC_SYNTHETICS_MQ_LOWERCASE)) {
+    headers->synthetics = nro_get_string(val, NULL);
+  } else if (0 == nr_strcmp(key_lc, X_NEWRELIC_TRANSACTION_MQ_LOWERCASE)) {
+    headers->transaction = nro_get_string(val, NULL);
+  } else if (0 == nr_strcmp(key_lc, X_NEWRELIC_DT_PAYLOAD_MQ_LOWERCASE)) {
+    headers->dt_payload = nro_get_string(val, NULL);
+  } else if (0 == nr_strcmp(key_lc, X_NEWRELIC_W3C_TRACEPARENT_MQ_LOWERCASE)) {
+    headers->traceparent = nro_get_string(val, NULL);
+  } else if (0 == nr_strcmp(key_lc, X_NEWRELIC_W3C_TRACESTATE_MQ_LOWERCASE)) {
+    headers->tracestate = nro_get_string(val, NULL);
+  }
+
+  nr_free(key_lc);
+  return NR_SUCCESS;
+}
 
+/*
+ * Purpose : Parse a Laravel 4.1+ job object for CATMQ metadata and update the
+ *           transaction type accordingly.
+ *
+ * Params  : 1. The job object.
+ */
+static void nr_laravel_queue_set_cat_txn(zval* job TSRMLS_DC) {
+  zval* json = NULL;
+  nrobj_t* payload = NULL;
+  nr_laravel_queue_headers_t headers = {NULL, NULL, NULL, NULL};
+
+  /*
+   * We're not interested in SyncJob instances, since they don't run in a
+   * separate queue worker and hence don't need to be linked via CATMQ.
+   */
+  if (nr_laravel_queue_is_sync_job(job TSRMLS_CC)) {
+    return;
+  }
+
+  /*
+   * Let's see if we can access the payload.
+   */
+  if (!nr_php_object_has_method(job, "getRawBody" TSRMLS_CC)) {
+    return;
+  }
+
+  json = nr_php_call(job, "getRawBody");
+  if (!nr_php_is_zval_non_empty_string(json)) {
+    goto end;
+  }
+
+  /*
+   * We've got it. Let's decode the payload and extract our CATMQ properties.
+   *
+   * Our nro code doesn't handle NULLs particularly gracefully, but it doesn't
+   * matter here, as we're not turning this back into JSON and aren't
+   * interested in the properties that could include NULLs.
+   */
+  payload = nro_create_from_json(Z_STRVAL_P(json));
+  nro_iteratehash(payload, (nrhashiter_t)nr_laravel_queue_iterate_headers,
+                  (void*)&headers);
+
+  if (headers.id && headers.transaction) {
+    nr_header_set_cat_txn(NRPRG(txn), headers.id, headers.transaction);
+  }
+
+  if (headers.synthetics) {
+    nr_header_set_synthetics_txn(NRPRG(txn), headers.synthetics);
+  }
+
+  if (headers.dt_payload || headers.traceparent) {
+    char* dt_payload = nr_strdup(headers.dt_payload);
+    char* tracestate = nr_strdup(headers.tracestate);
+    char* traceparent = nr_strdup(headers.traceparent);
+
+    nr_hashmap_t* header_map = nr_header_create_distributed_trace_map(
+        dt_payload, traceparent, tracestate);
```

Comment:
nr_header_create_distributed_trace_map already calls a [strdup](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/nr_header.c#L42), so we shouldn't need to do it here.

Reply:
Good call.
The reason to stdrup would be if we are storing a zval value that might go away, but that doesn't seem to be the case here.  I followed the preceeding and following calls and it does look like we strdup it right away, so we can claw back a little bit of performance in the dupping and the freeing.
Also updated the struct init at the top of the function to null ALL the values....

Please see:
[75e61fa](https://github.com/newrelic/newrelic-php-agent/pull/746/commits/75e61fafdf3bd689d3b87b21b3b06632f5311e27)